### PR TITLE
[WIP] External Cluster Integration

### PR DESCRIPTION
**Pull request by jtwinrelevanceai**

### CODE REVIEW

File: relevanceai/base.py [hunk 1]
```diff
@@ -16,3 +18,19 @@ def __init__(self, project: str, api_key: str):
...
+    @abstractmethod
+    def fit_transform(self, vectors):
+        raise NotImplementedError
+
+    @abstractmethod
+    def get_centroids(self):
+        raise NotImplementedError
+
+    def metadata(self):
+        return self.clusterer.__dict__
```
Comment (boba-and-beer) [inline relevanceai/base.py:36]: This requires a `hasattr`. If it does not have the `__dict__` attribute, this should not return the right thing.

File: relevanceai/dataset_api/dataset.py [hunk 2]
```diff
@@ -143,6 +146,24 @@ def bulk_fn(documents):
...
+        ---------
+        None
+
+        Returns
+        -------
+        vectors: np.ndarray
+            an array/matrix of all numeric values selected
+        """
+        documents = self.get_all_documents(self.dataset_id, select_fields=[self.field])
+        vectors = [np.array(sample[self.field]) for sample in documents]
```
Comment (boba-and-beer) [inline relevanceai/dataset_api/dataset.py:163]: where did `sample` come from?

File: relevanceai/dataset_api/dataset.py [hunk 3]
```diff
@@ -554,6 +575,20 @@ def cat_fields(documents, field_name):
             self.dataset_id, cat_fields, updating_args={"field_name": vector_name}
         )
 
+    def set_cluster_labels(self, vector_field, alias, labels):
```
Comment (boba-and-beer) [inline relevanceai/dataset_api/dataset.py:578]: `vector_fields`

File: relevanceai/dataset_api/dataset.py [hunk 4]
```diff
@@ -554,6 +575,20 @@ def cat_fields(documents, field_name):
...
+        def add_cluster_labels(documents):
+            docs = self.get_all_documents(self.dataset_id)
+            docs = list(filter(DocUtils.list_doc_fields, docs))
+            set_cluster_field = f"_cluster_.{vector_field}.{alias}"
```
Comment (boba-and-beer) [inline relevanceai/dataset_api/dataset.py:582]:
```
`"_cluster_" +  "." .join(vector_fields).lower() + "." + alias
```

File: relevanceai/dataset_api/dataset.py [hunk 5]
```diff
@@ -554,6 +575,20 @@ def cat_fields(documents, field_name):
...
+            self.set_field_across_documents(
+                set_cluster_field,
+                [f"cluster-{label}" for label in list(labels)],
```
Comment (boba-and-beer) [inline relevanceai/dataset_api/dataset.py:585]: I think we have a 
```
    def _label_clusters(self, labels):
        return [self._label_cluster(x) for x in labels]
```

So you should use: 

```
self.label_cluster
```

File: scripts/External Integration Clustering/case2_clustering.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
+from relevanceai import Client
```
Comment (boba-and-beer) [inline scripts/External Integration Clustering/case2_clustering.py:1]: Add quick comments at the top to describe what is going to happen in this file please :)

File: relevanceai/base.py [hunk 7]
```diff
@@ -16,3 +18,19 @@ def __init__(self, project: str, api_key: str):
         self.config = CONFIG
         # Initialize logger
         super().__init__()
+
+
+class ClusterBase:
```
Comment (boba-and-beer) [inline relevanceai/base.py:23]: We need this to be inheriting `ABC` or it won't work

**Review by boba-and-beer — CHANGES_REQUESTED** [review]
Basic structure is right. 
Minor changes going forward:  
- Make your scripts into actual scripts so that they can be run from command line (use `argparse`) 
- Add comments at the top of each script explaining what it does
- If certain things are missing imports, make sure to use them accordingly 
- `fit_transform` and `set_cluster_labels` should be wrapped into 1 function called `fit_dataset`

File: relevanceai/dataset_api/dataset.py (hunk 5 above)
Comment (jtwinrelevanceai) [reply relevanceai/dataset_api/dataset.py:585]: isnt this embedded inside the cluster class?

File: relevanceai/dataset_api/dataset.py (hunk 2 above)
Comment (jtwinrelevanceai) [reply relevanceai/dataset_api/dataset.py:163]: sample should written as documents here

**Comment by boba-and-beer** [comment]
Merged into feature/edit_docs

**Pull request closed without merging**